src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -323,3 +387,659 @@ X509* CryptoNative_X509UpRef(X509* x509)
+                        if (key != NULL)
+                        {
+                            EVP_PKEY_free(key);
+                        }
+
+                        if (cert == NULL && sk_X509_num(tmpStack) > 0)
+                        {
+                            cert = sk_X509_value(tmpStack, 0);
+                            X509_up_ref(cert);
+                        }
+                    }
+
+                    X509* popTmp;
+                    while ((popTmp = sk_X509_pop(tmpStack)) != NULL)
+                    {
+                        X509_free(popTmp);
+                    }
+
+                    PKCS12_free(p12);
+
+                    if (cert != NULL)
+                    {
+                        return cert;
+                    }
+                }
+
+                fclose(fp);
+            }
+        }
+    }
+
+    return NULL;
+}
+
+X509_STORE* CryptoNative_X509ChainNew(X509Stack* systemTrust, const char* userTrustPath)
+{
+    X509_STORE* store = X509_STORE_new();
+
+    if (store == NULL)
+    {
+        return NULL;
+    }
+
+    if (systemTrust != NULL)
+    {
+        int count = sk_X509_num(systemTrust);
+
+        for (int i = 0; i < count; i++)
+        {
+            if (!X509_STORE_add_cert(store, sk_X509_value(systemTrust, i)))
+            {
+                X509_STORE_free(store);
+                return NULL;
+            }
+        }
+    }
+
+    if (userTrustPath != NULL)
+    {
+        char* pathTmp;
+        char* nextFileWrite;
+        DIR* trustDir = OpenUserStore(userTrustPath, &pathTmp, &nextFileWrite);
+
+        if (trustDir != NULL)
+        {
+            X509* cert;
+            X509Stack* tmpStack = sk_X509_new_null();
+
+            while ((cert = ReadNextPublicCert(trustDir, tmpStack, pathTmp, nextFileWrite)) != NULL)
+            {
+                // cert refcount is 1
+                if (!X509_STORE_add_cert(store, cert))
+                {
+                    // cert refcount is still 1
+                    if (ERR_get_error() != ERR_PACK(ERR_LIB_X509, X509_F_X509_STORE_ADD_CERT, X509_R_CERT_ALREADY_IN_HASH_TABLE))
+                    {
+                        // cert refcount goes to 0
+                        X509_free(cert);
+                        X509_STORE_free(store);
+                        store = NULL;
+                        break;
+                    }
+                }
+
+                // if add_cert succeeded, reduce refcount to 1
+                // if add_cert failed (duplicate add), reduce refcount to 0
+                X509_free(cert);
+            }
+
+            sk_X509_free(tmpStack);
+            free(pathTmp);
+            closedir(trustDir);
+
+            if (store == NULL)
+            {
+                return NULL;
+            }
+
+            // PKCS12_parse can cause spurious errors.
+            // d2i_PKCS12_fp may have failed for invalid files.
+            // Just clear it all.
+            ERR_clear_error();
```

Comment:
We don't need to do this if store is NULL?

Reply:
No, if store is null we want to the error to be reported.  I'll add a comment :smile: